stistools/add_s_region.py

```diff
@@ -0,0 +1,403 @@
+                else:
+                    log.info('Dry-run - no changes made to science header')
+    return
+
+def write_keyword_to_header(extension, s_region):
+    """Write the S_REGION keyword to the header, creating the keyword if it
+    doesn't exist"""
+
+    if extension.header.get('S_REGION'):
+        # Keyword exists
+        extension.header['S_REGION'] = s_region
+        log.info('Existing S_REGION keyword updated')
+    else:
+        # Keyword doesn't exist, need to add it
+        extension.header.set('S_REGION', s_region, 'Spatial extent of the observation',
+        after='PA_APER')
+        log.info('New S_REGION keyword added and updated')
+    return
+
+def get_siaf_entry(h, aperture, detector):
+    """The SIAF entry aperture name doesn't correspond to the STIS
+    APERTURE keyword value.  Construct the entry from the aperture and
+    detector (here PROPAPER is used as the aperture keyword)
+
+    """
+    # All entries start with uppercase o
+    entry = 'O'
+    # Second letter depends on detector
+    detector_letters = {"CCD": "V",
+                        "NUV-MAMA": "N",
+                        "FUV-MAMA": "F"}
+    entry = entry + detector_letters[detector]
+    # The rest depends on the aperture
+    entry = entry + STIS_APERTURE_LOOKUP[aperture]
+    return h[entry]
+
+def get_wcs_limits(f1):
+    # For imaging data, any subarray will sometimes limit the observable
+    # footprint.  Calculate the limits of the field of view using the WCS,
+    # this will be compared with the limits from projecting the aperture
+    hdr1 = f1[1].header
+    crpix1 = hdr1['CRPIX1']
+    crpix2 = hdr1['CRPIX2']
+    # Calculate pixel scales in arcsec/pixel in the X (row) and Y (column)
+    # direction
+    cdelt1, cdelt2 = get_pixel_scales(f1)
+    nx = hdr1['NAXIS1']
+    ny = hdr1['NAXIS2']
+    xmin = cdelt1 * crpix1 * -1.0
+    xmax = cdelt1 * (nx - crpix1)
+    ymin = cdelt2 * crpix2 * -1.0
+    ymax = cdelt2 * (ny - crpix2)
+    # If the x-axis is dispersed, don't use the WCS limits
+    if hdr1['CTYPE1'] == 'WAVE':
+        xmin = None
+        xmax = None
+    return (xmin, xmax, ymin, ymax)
+
+def get_pixel_scales(f1):
+    """Calculate the pixel scales in the X (row) and Y (column) directions
+    from the WCS CD matrix, in arcsec/pixel
+    """
+    hdr = f1[1].header
+    cd1_1 = hdr['CD1_1']
+    cd1_2 = hdr['CD1_2']
+    cd2_1 = hdr['CD2_1']
+    cd2_2 = hdr['CD2_2']
+    cdelt1 = math.sqrt(cd1_1*cd1_1 + cd2_1*cd2_1)
+    cdelt2 = math.sqrt(cd1_2*cd1_2 + cd2_2*cd2_2)
+    cdelt1 = cdelt1 * 3600.0
+    cdelt2 = cdelt2 * 3600.0
+    return cdelt1, cdelt2
+
+def get_siaf_limits(x, y):
+    """Given arrays of X and Y values from the SIAF in ideal coordinates
+    (relative to the aperture reference point), return the X and Y limits
+    """
+    xmin = x.min()
+    xmax = x.max()
+    ymin = y.min()
+    ymax = y.max()
+    return (xmin, xmax, ymin, ymax)
+
+def smallest_size(wcslimits, siaflimits):
```

Comment:
Maybe name this "get_smallest_aperture_corners" instead of "smallest_size"?

Reply:
It's not really the smallest aperture corners, it really is the smallest size in each dimension.  Depends on whether the aperture is larger than the pixel array in each dimension